control-plane/agents/core/src/core/task_poller.rs

```diff
@@ -0,0 +1,166 @@
+    /// (the counter is auto-reloaded)
+    pub(crate) fn poll(&mut self) -> bool {
+        if self.ready() {
+            self.reload();
+            true
+        } else {
+            self.decrement();
+            false
+        }
+    }
+    fn ready(&self) -> bool {
+        self.counter <= 1
+    }
+    fn decrement(&mut self) {
+        self.counter -= 1;
+    }
+    fn reload(&mut self) {
+        self.counter = self.period;
+    }
+}
+
+/// Poll Context passed around the poll handlers
+pub(crate) struct PollContext {
+    /// Event that triggered this poll
+    event: PollEvent,
+    /// Core Registry
+    registry: Registry,
+}
+impl PollContext {
+    /// Create a context for a `PollEvent` with the global `Registry`
+    pub(crate) fn from(event: &PollEvent, registry: &Registry) -> Self {
+        Self {
+            event: event.clone(),
+            registry: registry.clone(),
+        }
+    }
+    /// Get a reference to the core registry
+    pub(crate) fn registry(&self) -> &Registry {
+        &self.registry
+    }
+
+    #[allow(dead_code)]
+    /// Get a reference to the event that triggered this poll
+    pub(crate) fn event(&self) -> &PollEvent {
+        &self.event
+    }
+}
+
+/// Trait used by all reconciliation loops
+#[async_trait::async_trait]
+pub(crate) trait TaskPoller: Send + Sync + std::fmt::Debug {
+    /// Attempts to poll this poller, which will poll itself depending on the `PollEvent`
+    #[tracing::instrument(skip(context), err)]
+    async fn try_poll(&mut self, context: &PollContext) -> PollResult {
+        tracing::trace!("Entering trace call");
+        let result = if self.poll_ready(context).await {
+            self.poll(context).await
+        } else {
+            PollResult::Ok(PollerState::Idle)
+        };
+        tracing::trace!("Leaving trace call");
+        result
+    }
+
+    /// Force poll the poller
+    async fn poll(&mut self, context: &PollContext) -> PollResult;
+
+    /// Polls the ready state and returns true if ready
+    async fn poll_ready(&mut self, context: &PollContext) -> bool {
+        match context.event() {
+            PollEvent::TimedRun => self.poll_counter(context).await,
+            _ => self.poll_event(context).await,
+        }
+    }
+
+    /// Polls the `PollCounter` and returns true if ready
+    async fn poll_counter(&mut self, _context: &PollContext) -> bool {
+        true
+    }
+
+    /// Determine if self should be polled for this `PollEvent`
+    async fn poll_event(&mut self, context: &PollContext) -> bool {
+        match context.event() {
+            PollEvent::TimedRun => true,
+            PollEvent::Triggered(_event) => true,
+            PollEvent::Shutdown => true,
+        }
+    }
+
+    /// Convert from a vector of results to a single result
+    fn squash_results(results: Vec<PollResult>) -> PollResult
+    where
+        Self: Sized,
+    {
+        let mut results = results.into_iter();
+        match results.find(|r| r.is_err()) {
```

Comment:
In the future we might want to reconsider what we do here. If multiple reconcile loops fail, we might want to know which ones failed and why.

Reply:
From a logging point of view we should know the error from the trace macros:
`#[tracing::instrument(skip(context), err)]`
But if we want to handle the different errors, then yes, we'd need some more checks further down the line.